provider/datagen/src/lib.rs

```diff
@@ -133,6 +133,49 @@ pub fn keys_from_file<R: Read>(file: R) -> std::io::Result<Vec<ResourceKey>> {
         .collect())
 }
 
+/// Parses a compiled binary and returns a list of used [`ResourceKey`]s. Unknown
+/// key names are ignored.
+///
+/// # Example
+///
+/// #### build.rs
+/// ```no_run
+/// # use icu_provider::ResourceMarker;
+/// # use std::fs::File;
+/// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+/// assert_eq!(
+///     icu_datagen::keys_from_bin("target/release/my-app")?,
+///     vec![
+///         icu_list::provider::AndListV1Marker::KEY,
+///         icu_list::provider::OrListV1Marker::KEY,
+///     ],
+/// );
+/// # Ok(())
+/// # }
+/// ```
+pub fn keys_from_bin<P: AsRef<Path>>(path: P) -> std::io::Result<Vec<ResourceKey>> {
+    let file = std::fs::read(path.as_ref())?;
+    let candidates = (0..file.len())
+        .filter_map(|i| {
+            if file[i..].starts_with(icu_provider::leading_tag!().as_bytes()) {
+                let i = i + icu_provider::leading_tag!().len();
+                for j in i..Ord::min(i + 100, file.len()) {
+                    // Assume some maximum key size to not make this quadratic
+                    if file[j..].starts_with(icu_provider::trailing_tag!().as_bytes()) {
+                        return Some(&file[i..j]);
+                    }
+                }
+            }
```

Comment:
Suggestion: This need not be a quadratic search. A simple linear-time algorithm is to have a single cursor that walks forward; sets a flag when it sees a leading tag; then, when it sees a trailing tag, if the leading tag bit is set, takes the slice between that leading tag and the trailing tag it just found, and sets the flag back to false again.

Reply:
It's not quadratic, 100 is a constant :)

Reply:
Ah you mean we don't have to limit keys to 100? Sure, but I think this is nice because it keeps the code very simple. The quadratic behaviour (/100 limit) will only happen if we analyze a binary with a missing trailing tag anyway.

Reply:
I think the linear-time algorithm is simple as well, maybe even simpler than the quadratic one. This is optional.